tests/phpunit/tests/option/option.php

```diff
@@ -312,4 +312,207 @@ public function data_option_autoloading() {
+
+		$this->assertFalse( $updated, 'update_option should not return true when values are loosely equal.' );
+		$this->assertSame( $num_queries, get_num_queries(), 'The number of database queries should not change.' );
+	}
+
+	/**
+	 * Data provider.
+	 *
+	 * @return array
+	 */
+	public function data_update_option_type_juggling() {
+		return array(
+			// Truthy.
+			array( '1', '1' ),
+			array( '1', 1 ),
+			array( '1', 1.0 ),
+			array( '1', true ),
+			array( 1, '1' ),
+			array( 1, 1 ),
+			array( 1, 1.0 ),
+			array( 1, true ),
+			array( 1.0, '1' ),
+			array( 1.0, 1 ),
+			array( 1.0, 1.0 ),
+			array( 1.0, true ),
+			array( true, '1' ),
+			array( true, 1 ),
+			array( true, 1 ),
+			array( true, true ),
+
+			// Falsey.
+			array( '0', '0' ),
+			array( '0', 0 ),
+			array( '0', 0.0 ),
+			array( '0', false ),
+			array( 0, '0' ),
+			array( 0, 0 ),
+			array( 0, 0.0 ),
+			array( 0, false ),
+			array( 0.0, '0' ),
+			array( 0.0, 0 ),
+			array( 0.0, 0.0 ),
+			array( 0.0, false ),
+			array( false, '0', false ),
+			array( false, 0, false ),
+			array( false, 0.0, false ),
+		);
+	}
+
+	/**
+	 * @ticket 22192
+	 *
+	 * @covers ::update_option
+	 *
+	 * @dataProvider data_update_option_type_falsey_values
+	 *
+	 * @param mixed $old_value One of the values to compare.
+	 * @param mixed $new_value The other value to compare.
+	 * @param int   $expected  The expected result.
+	 */
+	public function test_update_option_with_falsey_values( $old_value, $new_value, $expected ) {
+		add_option( 'foo', $old_value );
+		$num_queries = get_num_queries();
+
+		$updated = update_option( 'foo', $new_value );
+
+		$this->assertSame( $expected, $updated, 'update_option should not match expeted value when values are loosely equal.' );
+		$this->assertSame( 1, get_num_queries() - $num_queries, 'The number of database queries should not change.' );
+	}
+
+	/**
+	 * Data provider.
+	 *
+	 * @return array
+	 */
+	public function data_update_option_type_falsey_values() {
+		return array(
+			array( false, '', false ),
+			array( false, null, false ),
+			array( false, array(), true ),
+		);
+	}
+
+	/**
+	 * Tests that update_option() does not store an option
+	 * that uses an unfiltered default value of (bool) false.
+	 *
+	 * @ticket 22192
+	 *
+	 * @covers ::update_option
+	 */
+	public function test_update_option_should_not_store_option_with_default_value_false() {
+		global $wpdb;
+
+		$option = 'update_option_default_false';
+		update_option( $option, false );
+
+		$actual = $wpdb->query(
+			$wpdb->prepare(
+				"SELECT option_name FROM $wpdb->options WHERE option_name = %s LIMIT 1",
+				$option
+			)
+		);
+
+		$this->assertSame( 0, $actual );
+	}
+
+	/**
+	 * Tests that update_option() does not store an option
+	 * that uses a filtered default value.
+	 *
+	 * @ticket 22192
+	 *
+	 * @covers ::update_option
+	 */
+	public function test_update_option_should_not_store_option_with_filtered_default_value() {
```

Comment:
Same as the above, this test IMO tests the wrong thing. This _should_ add the option.